eng/pipelines/templates/jobs/smoke.tests.yml

```diff
@@ -179,36 +179,43 @@ jobs:
               return $line
             }
 
-            $dependencies | Out-File $env:REQUIREMENTS
+            $dependencies | Out-File $(requirements)
 
           displayName: Override requirements with pipeline build artifact versions
-          env:
-            REQUIREMENTS: $(requirements)
 
-        - pwsh: pip install -r "$env:REQUIREMENTS" --no-deps --upgrade
+        - pwsh: |
+            $ErrorActionPreference = "Continue"
+            while ($retries++ -lt 4) {
+              if ($retries -ge 4) {
+                $ErrorActionPreference = "Stop"
```

Comment:
How does this help? I was expecting to just have an exit below in `if ($LASTEXITCODE)` block that if the retry count ass been reached then `exit $LASTEXITCODE`

Reply:
So on the last loop, changing ErrorActionPreference back to stop, the next line that's going to get executed is the pip install, so if that fails, the pipeline would just throw an exception and terminate the job. So basically, if last pip install doesn't work, it exits.

Reply:
Given we are relying on LASTEXITCODE on all the other retries I would stick with that for the last call as well we don't want to change the behavior slightly between failure modes. We should fail the same way on each iteration.

Reply:
Ok makes sense